packages/actor-rdf-resolve-hypermedia-links-queue-wrapper-info-occupancy/lib/LinkQueueLogger.ts

```diff
@@ -0,0 +1,127 @@
+import type { ILinkQueue, ILink } from '@comunica/bus-rdf-resolve-hypermedia-links-queue';
+import { LinkQueueWrapper } from '@comunica/bus-rdf-resolve-hypermedia-links-queue';
+import type { Logger } from '@comunica/types';
+import { PRODUCED_BY_ACTOR } from '@comunica/types-link-traversal';
+import type { Algebra } from 'sparqlalgebrajs';
+
+/**
+ * The type of event
+ */
+export enum EventType {
+  Push,
+  Pop,
+}
+
+/**
+ * A link queue event
+ */
+interface ILinkQueueEvent {
+  type: EventType;
+  link: IURLStatistic;
+  query: Algebra.Operation;
+}
+/**
+ * Optional parameters necessitating special processing
+ */
+interface IOptionalLinkQueueParameters {
+}
+
+/**
+ * Information about an URL
+ */
+interface IURLStatistic {
+  url: string;
+  reachability_criteria: string | null;
+  timestamp?: number;
+  parent?: IURLStatistic;
+}
+/**
+ * A link queue that log information about event happening in the queue
+ */
+export class LinkQueueLogger extends LinkQueueWrapper {
+  public readonly query: Algebra.Operation;
+  private readonly logger: Logger;
+  public static readonly LINK_QUEUE_EVENT_NAME = '<Link queue occupancy>';
+  public static readonly LINK_QUEUE_DIDNT_STARTED_EMPTY_MESSAGE = 'the link queue didn\'t started empty';
+
+  /**
+   *
+   * @param {ILinkQueue & IOptionalLinkQueueParameters} linkQueue - The link queue with optional public parameters
+   * @param {Algebra.Operation} query - The current query
+   */
+  public constructor(linkQueue: ILinkQueue & IOptionalLinkQueueParameters, query: Algebra.Operation, logger: Logger) {
+    super(linkQueue);
+    this.logger = logger;
+
+    if (!linkQueue.isEmpty()) {
+      this.logger.warn(
+        LinkQueueLogger.LINK_QUEUE_EVENT_NAME,
+        {
+          message: LinkQueueLogger.LINK_QUEUE_DIDNT_STARTED_EMPTY_MESSAGE,
+        },
+      );
+    }
+    this.query = JSON.parse(JSON.stringify(query));
+
+    Object.freeze(this.query);
+  }
+
+  private static getReachability(link: ILink): string | null {
+    const metadata = link.metadata;
+    if (metadata !== undefined) {
+      return metadata[PRODUCED_BY_ACTOR]?.name ?? null;
+    }
+    return null;
+  }
+
+  public override push(link: ILink, parent: ILink): boolean {
+    const resp: boolean = super.push(link, parent);
+    if (resp) {
+      const parentStatisticLink: IURLStatistic = {
+        url: parent.url,
+        reachability_criteria: LinkQueueLogger.getReachability(parent),
+      };
+      const statisticLink: IURLStatistic = {
```

Comment:
Here it seems like you're not using all metadata possibly available. For example, in [extract links predicates](https://github.com/comunica/comunica-feature-link-traversal/pull/136/files#diff-1b5834f1eccb1aa4b05ad1a555829f95534fa2c58cd83199918058b91c5d6f33) you record what predicate was used to get the link, but you don't seem to include that in your logging (if I'm not missing anything). It could be interesting to include this.